features/shard: avoid repeatative calls to gf_uuid_unparse()

vh05 opened the pull request:
The issue is shard_make_block_abspath() calls gf_uuid_unparse()
every time while constructing shard path. The gfid can be parsed
and saved once and passed while constructing the path. Thus
we can avoid calling gf_uuid_unparse().

Fixes: #1423

Change-Id: Ia26fbd5f09e812bbad9e5715242f14143c013c9c
Signed-off-by: Vinayakswami Hariharmath <vharihar@redhat.com>

vh05 commented:
/run regression

vh05 commented on tests/bugs/shard/issue-1425.t:
```diff
@@ -21,7 +21,13 @@ TEST $CLI volume profile $V0 start
+
 TEST fallocate -l 20M $M0/foo
+
+# There should be 1+4 shards and we expect 4 lookups less than on the build without this patch
```
Not relevant to this patch. 

The reason for this change is that the test was failing at line 34 because of inconsistent lookup numbers. I am moving that check once the fallocate is done to minimize the window and I think that's what test 1425 also expects

xhernandez commented on xlators/features/shard/src/shard.c:
```diff
@@ -1064,23 +1064,26 @@ shard_common_resolve_shards(call_frame_t *frame, xlator_t *this,
     else
         gf_uuid_copy(gfid, local->base_gfid);
 
+    /* Build base shard path before appending index of the shard */
+    strcpy(path, "/" GF_SHARD_DIR "/");
+    uuid_utoa_r(gfid, path + sizeof(GF_SHARD_DIR) + 1);
+    int prefix_len = sizeof(GF_SHARD_DIR) + GF_UUID_BUF_SIZE;
+
```
This is used several times. Could you create a function to initialize this ?

vh05 replied:
Yes. Making it a macro

xhernandez commented on xlators/features/shard/src/shard.c:
```diff
@@ -1064,23 +1064,26 @@ shard_common_resolve_shards(call_frame_t *frame, xlator_t *this,
-
-        shard_make_block_abspath(shard_idx_iter, gfid, path, sizeof(path));
-
+        snprintf(path + prefix_len, sizeof(path) - prefix_len, ".%d",
+                 shard_idx_iter);
```
Not strictly necessary, but if you create a function to initialize the `path`, it would be cleaner to also create a function to do this.

vh05 replied:
Yep. Making it a macro

xhernandez commented on xlators/features/shard/src/shard.c:
```diff
@@ -1013,7 +1013,7 @@ shard_common_resolve_shards(call_frame_t *frame, xlator_t *this,
     int i = -1;
     uint32_t shard_idx_iter = 0;
-    char path[PATH_MAX] = {
+    char path[sizeof(GF_SHARD_DIR) + GF_UUID_BUF_SIZE + 16] = {
```
Defining this length as a constant would be better.

vh05 replied:
Thought this will bit informative. Yes. Making it a macro

vh05 committed:
features/shard: avoid repeatative calls to gf_uuid_unparse()

The issue is shard_make_block_abspath() calls gf_uuid_unparse()
every time while constructing shard path. The gfid can be parsed
and saved once and passed while constructing the path. Thus
we can avoid calling gf_uuid_unparse().

Fixes: #1423

Change-Id: I7cd5a73b84bd7da7a6eb7bb5f16f5872fe1d204c
Signed-off-by: Vinayakswami Hariharmath <vharihar@redhat.com>

vh05 commented:
/run regression

itisravi commented on tests/bugs/shard/issue-1425.t:
```diff
@@ -21,7 +21,13 @@ TEST $CLI volume profile $V0 start
 
 TEST $GFS --volfile-id=$V0 --volfile-server=$H0 $M0
 
+echo `$CLI volume profile $V0 info incremental | grep -w LOOKUP | awk '{print $8}'` > /dev/zero
```
Nit: You can also `$CLI volume profile $V0 info clear`.

vh05 replied:
sure

itisravi commented on xlators/features/shard/src/shard.c:
```diff
@@ -1064,23 +1067,23 @@ shard_common_resolve_shards(call_frame_t *frame, xlator_t *this,
-                         "present. gfid=%s. Saving inode for future.",
-                         shard_idx_iter, uuid_utoa(inode->gfid));
+                         "Shard %s already present. Saving inode for future.",
+                         path);
```
Just wondering, why remove gfid from the log message?

vh05 replied:
We were using uuid_utoa(inode->gfid) to get the gfid every time to print the gfid. Rather we can use the path of it which gives the gfid info and we get rid of calling uuid_utoa() every time.

itisravi replied:
Oh, I thought the gfid part that is present in the path is that of the base shard (and not the actual gfid of the shard which is different for each shard).

itisravi commented on xlators/features/shard/src/shard.c:
```diff
@@ -1064,23 +1067,23 @@ shard_common_resolve_shards(call_frame_t *frame, xlator_t *this,
             continue;
         }
-
-        shard_make_block_abspath(shard_idx_iter, gfid, path, sizeof(path));
```
Do we need shard_make_block_abspath() in the code base any more?

vh05 replied:
Not needed as of now. Thought to remove it but just that if somebody wants to use it. Anyway, I will remove it in the next patch update

VHariharmath-rh commented:
./run regression

VHariharmath-rh commented:
./run regression

itisravi commented:
/run regression

xhernandez commented on xlators/features/shard/src/shard.c:
```diff
@@ -1013,7 +1004,8 @@ shard_common_resolve_shards(call_frame_t *frame, xlator_t *this,
     uint32_t shard_idx_iter = 0;
-    char path[PATH_MAX] = {
+    int prefix_len = 0;
+    char path[SHARD_PATH_MAX] = {
```
Don't initialize it. It's a waste of time.

vh05 replied:
sure

xhernandez commented on xlators/features/shard/src/shard.h:
```diff
@@ -199,6 +199,19 @@ shard_unlock_entrylk(call_frame_t *frame, xlator_t *this);
+        prefix_len = sizeof(GF_SHARD_DIR) + GF_UUID_BUF_SIZE;                  \
+    } while (0)
+
+#define SHARD_BUILD_ABS_PATH(path, prefix_len, shard_idx_iter)                 \
```
Maybe a more meaningful name would be `SHARD_APPEND_IDX` or something similar. With the current name it seems to create a full path.

Could this be created as a function instead of a macro ?

xhernandez commented on xlators/features/shard/src/shard.h:
```diff
@@ -199,6 +199,19 @@ shard_unlock_entrylk(call_frame_t *frame, xlator_t *this);
+
+#define SHARD_BUILD_ABS_PATH(path, prefix_len, shard_idx_iter)                 \
+    do {                                                                       \
+        snprintf(path + prefix_len, sizeof(path) - prefix_len, ".%d",          \
```
Using `sizeof()` here is very dangerous if a caller uses a pointer instead of an array. It's better to pass the size as an argument.

vh05 replied:
sure

xhernandez commented on xlators/features/shard/src/shard.h:
```diff
@@ -199,6 +199,19 @@ shard_unlock_entrylk(call_frame_t *frame, xlator_t *this);
         }                                                                      \
     } while (0)
 
+#define SHARD_BUILD_BASE_PATH(path, gfid, prefix_len)                          \
```
Modifying macro arguments inside a macro is dangerous because it's not obvious to developers (arguments are not normally modified unless they are passed by reference). A better approach would be to return that value.

Another question is why this is not created as a function ? I don't see anything that requires a macro.

vh05 replied:
sure

VHariharmath-rh commented:
./run regression

mykaul commented on xlators/features/shard/src/shard.c:
```diff
@@ -16,6 +16,8 @@
 #include <glusterfs/defaults.h>
 #include <glusterfs/statedump.h>
 
+#define SHARD_PATH_MAX sizeof(GF_SHARD_DIR) + GF_UUID_BUF_SIZE + 16
```
nit: 16 is a magic number?

vh05 replied:
16 bytes to accommodate the shard-index.

mykaul commented on xlators/features/shard/src/shard.c:
```diff
@@ -49,15 +51,19 @@ shard_make_block_bname(int block_num, uuid_t gfid, char *buf, size_t len)
-    };
+    strcpy(path, "/" GF_SHARD_DIR "/");
+    uuid_utoa_r(gfid, path + sizeof(GF_SHARD_DIR) + 1);
+    return (sizeof(GF_SHARD_DIR) + GF_UUID_BUF_SIZE);
```
what about the "/" before and after GF_SHARD_DIR?

vh05 replied:
after the strcpy, the path would contain "/.shard/"

mykaul commented on xlators/features/shard/src/shard.c:
```diff
@@ -49,15 +51,19 @@ shard_make_block_bname(int block_num, uuid_t gfid, char *buf, size_t len)
-    gf_uuid_unparse(gfid, gfid_str);
-    snprintf(filepath, len, "/%s/%s.%d", GF_SHARD_DIR, gfid_str, block_num);
+static void
+shard_append_index(char *path, int path_size, int prefix_len,
```
nit: I'm not sure a function with single sprintf() statement is more readable than just putting that snprintf() directly.

vh05 replied:
Func shard_append_index() would give insight about what we are doing there. I will make it "static inline"

vh05 commented:
/recheck smoke

vh05 commented:
/recheck smoke

vh05 commented:
/recheck smoke

VHariharmath-rh commented:
./recheck smoke

vh05 commented:
/run regression

vh05 commented:
/recheck smoke

xhernandez commented on xlators/features/shard/src/shard.c:
```diff
@@ -16,6 +16,8 @@
 #include <glusterfs/defaults.h>
 #include <glusterfs/statedump.h>
 
+#define SHARD_PATH_MAX sizeof(GF_SHARD_DIR) + GF_UUID_BUF_SIZE + 16
```
To avoid problems, always put macro computations between parenthesis.

vh05 replied:
sure

vh05 committed:
features/shard: avoid repeatative calls to gf_uuid_unparse()

The issue is shard_make_block_abspath() calls gf_uuid_unparse()
every time while constructing shard path. The gfid can be parsed
and saved once and passed while constructing the path. Thus
we can avoid calling gf_uuid_unparse().

Fixes: #1423
Change-Id: I59931b86017d93e97b13906541ffde6aae51abae
Signed-off-by: Vinayakswami Hariharmath <vharihar@redhat.com>

vh05 commented:
/run regression

vh05 commented:
/recheck smoke

vh05 commented:
/run regression

xhernandez reviewed (approved):
It looks good to me. I'll wait for others to review before merging.

mykaul commented on xlators/features/shard/src/shard.c:
```diff
@@ -16,6 +16,8 @@
 #include <glusterfs/defaults.h>
 #include <glusterfs/statedump.h>
 
+#define SHARD_PATH_MAX (sizeof(GF_SHARD_DIR) + GF_UUID_BUF_SIZE + 16)
```
nit: 16 here looks like a magic number. If there's another round of this patch series, worth fixing (otherwise, let's not bother)

itisravi reviewed (approved):
+1
